hoodie-client/src/test/java/com/uber/hoodie/func/TestBufferedIterator.java

```diff
@@ -63,106 +75,203 @@ public void afterTest() {
+          while (itr.hasNext()) {
+            try {
+              buf.insertRecord(itr.next());
+            } catch (Exception e) {
+              throw new HoodieException(e);
+            }
+          }
+          return true;
+        }));
+      }
+    }
+
+    BufferProducer<HoodieRecord> compositeProducer =
+        new CompositeBufferProducer<HoodieRecord>(producers);
+
+    // Start Producer
+    Future<Void> res = compositeProducer.start(buffer, recordReader, () -> null);
+
+    // Used to ensure that consumer sees the records generated by a single producer in FIFO order
+    Map<Integer, Integer> lastSeenMap = IntStream.range(0, numProducers).boxed()
+        .collect(Collectors.toMap(Function.identity(), x -> -1));
+    Map<Integer, Integer> countMap = IntStream.range(0, numProducers).boxed()
+        .collect(Collectors.toMap(Function.identity(), x -> 0));
+
+    // Read recs and ensure we have covered all producer recs.
+    while (buffer.iterator().hasNext()) {
+      final HoodieRecordBufferedIteratorPayload payload = buffer.iterator().next();
+      final HoodieRecord rec = (HoodieRecord) payload.getInputPayload();
+      Tuple2<Integer, Integer> producerPos = keyToProducerAndIndexMap.get(rec.getRecordKey());
+      Integer lastSeenPos = lastSeenMap.get(producerPos._1());
+      countMap.put(producerPos._1(), countMap.get(producerPos._1()) + 1);
+      lastSeenMap.put(producerPos._1(), lastSeenPos + 1);
+      // Ensure we are seeing the next record generated
+      Assert.assertEquals(lastSeenPos + 1, producerPos._2().intValue());
+    }
+
+    for (int i = 0; i < numProducers; i++) {
+      // Ensure we have seen all the records for each producers
+      Assert.assertEquals(Integer.valueOf(numRecords), countMap.get(i));
+    }
+
+    //Ensure producers completed fine
+    res.get();
   }
 
   // Test to ensure that record buffering is throttled when we hit memory limit.
+  @SuppressWarnings("unchecked")
   @Test(timeout = 60000)
-  public void testMemoryLimitForBuffering() throws IOException, InterruptedException {
+  public void testMemoryLimitForBuffering() throws Exception {
     final int numRecords = 128;
     final List<HoodieRecord> hoodieRecords = hoodieTestDataGenerator.generateInserts(commitTime, numRecords);
     // maximum number of records to keep in memory.
     final int recordLimit = 5;
-    final long memoryLimitInBytes = recordLimit * SizeEstimator.estimate(hoodieRecords.get(0));
-    final BufferedIterator<HoodieRecord, AbstractBufferedIteratorPayload> bufferedIterator =
-        new BufferedIterator(hoodieRecords.iterator(), memoryLimitInBytes,
-        LazyInsertIterable.bufferedItrPayloadTransform(HoodieTestDataGenerator.avroSchema));
-    Future<Boolean> result = recordReader.submit(() -> {
-      bufferedIterator.startBuffering();
-      return true;
+    LazyInsertIterable.bufferedItrPayloadTransform(HoodieTestDataGenerator.avroSchema);
+    final long objSize = ObjectSizeCalculator.getObjectSize(
```

Comment:
Since we are changing this to use raw ObjectSizeCalculator, could we use a Converter implementation here instead ? Something that implements this : https://github.com/uber/hudi/blob/master/hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/converter/Converter.java

Reply:
As this is a unit-test,  I am assuming you are referring to using converter like interface for MemoryBoundedBuffer. It made sense for ExternalSpillableMap where clients need to control ser-de using converter. In the case of Size estimation, I am not sure why we are using multiple implementations ? If possible, I would like to see one implementation used consistently everywhere in hoodie. Is there any reason not to do that ?

Reply:
Yes, I'm also in favor of one implementation. The BufferConverter can just implement getSizeEstimate() and ignore other API's ? If we use the converter then what library/code is used for size estimation is isolated in one place and easier to change. Thoughts ?

Reply:
I have added a SizeEstimator interface with default implementation (using Object Estimator) and made Converter interface inherit it.